File: coroutine-example/src/main/java/structuredconcurrency/StructuredTaskScopeExample.java
//package structuredconcurrency;
//
//import java.util.concurrent.Callable;
//import java.util.concurrent.StructuredTaskScope;
//
//public class StructuredTaskScopeExample {
//
//    public static void main(String[] args) {
//        var example = new StructuredTaskScopeExample();
//        example.sendMessage();
//    }
//
//    public void sendMessage() {
//        try (var scope = new StructuredTaskScope.ShutdownOnFailure()) {
//            scope.fork((Callable<Void>) () -> {
//                sendEmail();
//                return null;
//            });
//
//            scope.fork((Callable<Void>) () -> {
//                sendSms();
//                return null;
//            });
//
//            scope.join();
//            scope.throwIfFailed((e) -> new RuntimeException("Failed to send"));
//
//        } catch (InterruptedException e) {
//            throw new RuntimeException(e);
//        }
//    }
//
//    public static void sendEmail() {
//        System.out.println("이메일 발송 완료");
//    }
//
//    public static void sendSms() {
//        System.out.println("SMS 발송 완료");
//    }
//}
//
//

File: coroutine-example/src/main/kotlin/virtualthread/structuredTaskScopeExample.kt
package virtualthread

import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import java.util.concurrent.StructuredTaskScope
import java.util.concurrent.StructuredTaskScope.Subtask


fun main() {
    getUsers1().forEach {
        println(it)
    }
}

fun getUsers1(): List<UserResponse> {
    StructuredTaskScope.ShutdownOnFailure().use { scope ->
        val fork1: Subtask<UserResponse> = scope.fork { callApi(1) }
        val fork2: Subtask<UserResponse> = scope.fork { callApi(2) }

        scope.join()
        scope.throwIfFailed { e ->
            RuntimeException("Failed to get user", e)
        }

        return listOf(fork1.get(), fork2.get())
    }
}

suspend fun getUsers(): List<UserResponse> = coroutineScope {
    val user1 = async { callApi(1) }
    val user2 = async { callApi(2) }

    listOf(user1.await(), user2.await())
}
//
//fun getUsers() = runBlocking {
//    val user1 = async { callApi(1) }
//    val user2 = async { callApi(2) }
//
//    listOf(user1.await(), user2.await())
//}





fun callApi(id: Long): UserResponse {
    val name = if (id == 1L) {
        "아이언맨"
    } else {
        throw RuntimeException("User not found")
    }
    return UserResponse(id, name)
}

data class UserResponse(
    val id: Long,
    val name: String,
)